### ISSUE TITLE
Proposal ref(builder): do not build slugbuilder and slugrunner locally

### ISSUE THREAD

**Issue by aledbf** [deis/deis#3395]
It could possible to interact with ceph (S3) and just download pre-built files containing the `slugbuilder` and `slugrunner` images and add those files in the `deis-registry` bucket.

This will avoid the i/o issues building the images, reduce the `deis-builder` size and optionally it could be possible to download the images and host that from a local server.

This makes sense?

**Comment by carmstrong** [comment]
@aledbf I'm not sure relying on Ceph is an advantage here, but I can see the motivation behind the request. We're going to look at optimizing builder for the next release. Thanks for bringing it up!

**Comment by bacongobbler** [comment]
So here's what I explained in the meeting yesterday about why this is necessary:

In previous versions, you had to specify an etcd key to pull in both slugrunner and slugbuilder into the builder. However as we all have known by now, `docker pull` requires network bandwidth, which in some cases - like when you're running a local vagrant cluster or DockerHub is down - can be slow at times. Therefore, building the images themselves directly inside builder actually sped up the boot process rather than slowed it down. It also made hacking on the builder a lot easier; one would just modify the source, re-compile builder and their changes were available.

In short, I think it's a *great* thing that it's being built inside the builder. We're looking to improve the builder boot pipeline, however. #3453 is just one step in the right direction, IMO (pushing slugrunner to the registry is probably the slowest operation the builder's doing right now).

**Comment by aledbf** [comment]
> probably the slowest operation the builder's doing right now

Exactly. Why not just populate the registry from a tgz with an "oficial" image?

**Comment by bacongobbler** [comment]
Because that breaks one of the biggest tenets of Deis: hackability (new word yay!). Hacking on Deis should be comfortable and easy, rather than having to set up your own remote S3 endpoint just to host your own custom slug* images.

**Comment by bacongobbler** [comment]
cedarish is an excuse because it's a complete mimic of heroku's cedar stack. If you're modifying it then you're doing something completely out of whack :P

**Comment by aledbf** [comment]
I just don't want to build `slugrunner` and `slugbuilder` but I don't want the old behavior either (docker pull for both images). The current weight of deis-builder is at least 370MB because of the cedarish tgz that every time you setup a cluster you need to download from docker hub.

**Comment by bacongobbler** [comment]
Slugrunner and slugbuilder need to be in the builder in order to compile applications that rely on Heroku buildpacks, so there's no way around it *not* being in the builder. There's a few requirements if we're going forward with this:

 - slugrunner and slugbuilder must be present in the builder
 - users need to be able to fork slugrunner/slugbuilder for their own changes

I don't see how moving the layers from the builder to the registry (backed by ceph) will optimize what we currently have, but I'm open to suggestions.

**Comment by bacongobbler** [comment]
this has been partially improved with #3933 but they're still built on initial boot. After that they're cached and won't be rebuilt again. However, if you attach a deis-builder-data volume with slug* pre-cached then that should resolve your issue. Does that seem fair, @aledbf?

**Comment by aledbf** [comment]
@bacongobbler yes, closing.
